File: Services/FirestoreService.swift
//
//  FirestoreService.swift
//  Shaglni
//
//  Compatibility shim that forwards every legacy callback-based API to the new
//  async repositories. New code should use the repositories directly (they expose
//  `@Published` arrays backed by live Firestore listeners — no polling needed).
//

import Foundation
import FirebaseFirestore

@MainActor
final class FirestoreService {
    static let shared = FirestoreService()
    private init() {}

    private var jobs:        JobsRepository        { .shared }
    private var offers:      OffersRepository      { .shared }
    private var contractors: ContractorsRepository { .shared }
    private var portfolio:   PortfolioRepository   { .shared }

    // MARK: - Jobs

    func createJob(_ job: Job, completion: @escaping (Result<String, Error>) -> Void) {
        Task {
            do { completion(.success(try await jobs.create(job))) }
            catch { completion(.failure(error)) }
        }
    }

    func fetchJobs(status: JobStatus? = nil, completion: @escaping (Result<[Job], Error>) -> Void) {
        Task {
            do {
                var query: Query = Firestore.firestore().collection("jobs")
                if let status { query = query.whereField("status", isEqualTo: status.rawValue) }
                let snap = try await query.order(by: "datePosted", descending: true)
                    .limit(to: JobsRepository.queryLimit).getDocuments()
                completion(.success(snap.decoded(as: Job.self)))
            } catch {
                completion(.failure(error))
            }
        }
    }

    func fetchJobsByUser(userId: String, completion: @escaping (Result<[Job], Error>) -> Void) {
        Task {
            do {
                let snap = try await Firestore.firestore().collection("jobs")
                    .whereField("createdBy", isEqualTo: userId)
                    .order(by: "datePosted", descending: true)
                    .limit(to: JobsRepository.queryLimit)
                    .getDocuments()
                completion(.success(snap.decoded(as: Job.self)))
            } catch {
                completion(.failure(error))
            }
        }
    }

    func fetchJob(jobId: String, completion: @escaping (Result<Job, Error>) -> Void) {
        Task {
            do { completion(.success(try await jobs.fetch(jobId: jobId))) }
            catch { completion(.failure(error)) }
        }
    }

    func deleteJob(jobId: String, completion: @escaping (Result<Void, Error>) -> Void) {
        Task {
            do { try await jobs.delete(jobId: jobId); completion(.success(())) }
            catch { completion(.failure(error)) }
        }
    }

    func updateJobStatus(jobId: String, status: JobStatus, completion: @escaping (Result<Void, Error>) -> Void) {
        Task {
            do { try await jobs.updateStatus(jobId: jobId, status: status); completion(.success(())) }
            catch { completion(.failure(error)) }
        }
    }

    // MARK: - Offers

    func submitOffer(_ offer: Offer, jobId: String, completion: @escaping (Result<String, Error>) -> Void) {
        Task {
            do { completion(.success(try await offers.submit(offer, jobId: jobId))) }
            catch { completion(.failure(error)) }
        }
    }

    func fetchOffersForJob(jobId: String, completion: @escaping (Result<[Offer], Error>) -> Void) {
        Task {
            do { completion(.success(try await offers.fetchOffers(jobId: jobId))) }
            catch { completion(.failure(error)) }
        }
    }

    func fetchOffersByContractor(contractorId: String, completion: @escaping (Result<[Offer], Error>) -> Void) {
        Task {
            do {
                let snap = try await Firestore.firestore().collectionGroup("offers")
                    .whereField("contractorId", isEqualTo: contractorId)
                    .order(by: "createdAt", descending: true)
                    .limit(to: JobsRepository.queryLimit)
                    .getDocuments()
                completion(.success(snap.decoded(as: Offer.self)))
            } catch { completion(.failure(error)) }
        }
    }

    func fetchOffersForJobPoster(userId: String, completion: @escaping (Result<[OfferWithJob], Error>) -> Void) {
        Task {
            do {
                let myJobsSnap = try await Firestore.firestore().collection("jobs")
                    .whereField("createdBy", isEqualTo: userId)
                    .order(by: "datePosted", descending: true)
                    .limit(to: JobsRepository.queryLimit)
                    .getDocuments()
                let myJobs = myJobsSnap.decoded(as: Job.self)
                completion(.success(try await offers.fetchOffersForJobPoster(userId, jobs: myJobs)))
            } catch { completion(.failure(error)) }
        }
    }

    func updateOfferStatus(jobId: String, offerId: String, status: OfferStatus, completion: @escaping (Result<Void, Error>) -> Void) {
        Task {
            do { try await offers.updateStatus(jobId: jobId, offerId: offerId, status: status); completion(.success(())) }
            catch { completion(.failure(error)) }
        }
    }

    func sendCounterOffer(jobId: String, offerId: String, counterPrice: Double, message: String, completion: @escaping (Result<Void, Error>) -> Void) {
        Task {
            do { try await offers.sendCounterOffer(jobId: jobId, offerId: offerId, counterPrice: counterPrice, message: message); completion(.success(())) }
            catch { completion(.failure(error)) }
        }
    }

    func respondToCounterOffer(jobId: String, offerId: String, accept: Bool, counterPrice: Double?, completion: @escaping (Result<Void, Error>) -> Void) {
        Task {
            do {
                if accept {
                    guard let counterPrice, counterPrice > 0 else {
                        throw AppError.validation("Counter offer has no valid price")
                    }
                    try await offers.contractorAcceptsCounter(jobId: jobId, offerId: offerId, finalPrice: counterPrice)
                } else {
                    try await offers.contractorDeclinesCounter(jobId: jobId, offerId: offerId)
                }
                completion(.success(()))
            } catch { completion(.failure(error)) }
        }
    }

    // MARK: - Contractor Profiles

    func fetchContractorProfile(userId: String, completion: @escaping (Result<ContractorProfile, Error>) -> Void) {
        Task {
            do { completion(.success(try await contractors.fetchProfile(userId: userId))) }
            catch { completion(.failure(error)) }
        }
    }

    func fetchAllContractors(completion: @escaping (Result<[ContractorProfile], Error>) -> Void) {
        Task {
            do {
                let snap = try await Firestore.firestore().collection("contractorProfiles")
                    .order(by: "completedJobsCount", descending: true)
                    .limit(to: JobsRepository.queryLimit)
                    .getDocuments()
                completion(.success(snap.decoded(as: ContractorProfile.self)))
            } catch { completion(.failure(error)) }
        }
    }

    func updateContractorProfile(_ profile: ContractorProfile, completion: @escaping (Result<Void, Error>) -> Void) {
        Task {
            do { try await contractors.upsertProfile(profile); completion(.success(())) }
            catch { completion(.failure(error)) }
        }
    }

    // MARK: - Completed Jobs

    func addCompletedJob(_ completedJob: CompletedJob, completion: @escaping (Result<String, Error>) -> Void) {
        Task {
            do { completion(.success(try await portfolio.add(completedJob))) }
            catch { completion(.failure(error)) }
        }
    }

    func incrementCompletedJobsCount(contractorId: String, completion: @escaping (Result<Void, Error>) -> Void) {
        Task {
            do { try await contractors.incrementCompletedJobsCount(contractorId: contractorId); completion(.success(())) }
            catch { completion(.failure(error)) }
        }
    }

    func fetchCompletedJobs(contractorId: String, completion: @escaping (Result<[CompletedJob], Error>) -> Void) {
        Task {
            do { completion(.success(try await portfolio.fetchPortfolio(contractorId: contractorId))) }
            catch { completion(.failure(error)) }
        }
    }

    func deleteCompletedJob(_ completedJob: CompletedJob, completion: @escaping (Result<Void, Error>) -> Void) {
        Task {
            do { try await portfolio.delete(completedJob); completion(.success(())) }
            catch { completion(.failure(error)) }
        }
    }

    func fetchJobsWithAcceptedOffer(contractorId: String, completion: @escaping (Result<[JobWithOffer], Error>) -> Void) {
        Task {
            do {
                let snap = try await Firestore.firestore().collection("jobs")
                    .whereField("acceptedContractorId", isEqualTo: contractorId)
                    .order(by: "acceptedAt", descending: true)
                    .limit(to: JobsRepository.queryLimit)
                    .getDocuments()
                let jobs = snap.decoded(as: Job.self)
                var out: [JobWithOffer] = []
                for j in jobs {
                    guard let jid = j.id, let oid = j.acceptedOfferId else { continue }
                    let offerDoc = try await Firestore.firestore().collection("jobs").document(jid).collection("offers").document(oid).getDocument()
                    if let off = try? offerDoc.decode(as: Offer.self) {
                        out.append(JobWithOffer(job: j, offer: off))
                    }
                }
                completion(.success(out))
            } catch { completion(.failure(error)) }
        }
    }
}

File: android/app/src/main/java/com/shaglni/app/data/FirestoreService.kt
package com.shaglni.app.data

import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import com.google.firebase.firestore.ListenerRegistration
import com.shaglni.app.data.model.CompletedJob
import com.shaglni.app.data.model.ContractorProfile
import com.shaglni.app.data.model.Job
import com.shaglni.app.data.model.Offer
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.tasks.await
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class FirestoreService @Inject constructor() {
    private val db: FirebaseFirestore = FirebaseFirestore.getInstance()

    // Jobs
    suspend fun createJob(job: Job): String {
        val docRef = db.collection("jobs").add(job).await()
        return docRef.id
    }

    suspend fun fetchJobs(status: String? = null): List<Job> {
        var query: Query = db.collection("jobs")
        if (status != null) query = query.whereEqualTo("status", status)
        val snapshot = query.orderBy("datePosted", Query.Direction.DESCENDING).get().await()
        return snapshot.documents.mapNotNull { it.toObject(Job::class.java)?.copy(id = it.id) }
    }

    suspend fun fetchJobsByUser(userId: String): List<Job> {
        val snapshot = db.collection("jobs")
            .whereEqualTo("createdBy", userId)
            .orderBy("datePosted", Query.Direction.DESCENDING)
            .get().await()
        return snapshot.documents.mapNotNull { it.toObject(Job::class.java)?.copy(id = it.id) }
    }

    suspend fun fetchJob(jobId: String): Job? {
        val doc = db.collection("jobs").document(jobId).get().await()
        return if (doc.exists()) doc.toObject(Job::class.java)?.copy(id = doc.id) else null
    }

    suspend fun updateJob(job: Job) {
        require(job.id.isNotBlank()) { "Job ID is required for updates" }
        db.collection("jobs").document(job.id).set(job).await()
    }

    suspend fun deleteJob(jobId: String) {
        db.collection("jobs").document(jobId).delete().await()
    }

    // Offers
    suspend fun submitOffer(offer: Offer, jobId: String): String {
        val docRef = db.collection("jobs").document(jobId)
            .collection("offers").add(offer).await()
        return docRef.id
    }

    suspend fun fetchOffersForJob(jobId: String): List<Offer> {
        val snapshot = db.collection("jobs").document(jobId)
            .collection("offers")
            .orderBy("createdAt", Query.Direction.DESCENDING)
            .get().await()
        return snapshot.documents.mapNotNull { it.toObject(Offer::class.java)?.copy(id = it.id, jobId = jobId) }
    }

    suspend fun fetchOffersByContractor(contractorId: String): List<Offer> {
        val snapshot = db.collectionGroup("offers")
            .whereEqualTo("contractorId", contractorId)
            .orderBy("createdAt", Query.Direction.DESCENDING)
            .get().await()
        return snapshot.documents.mapNotNull {
            val jobId = it.reference.parent.parent?.id ?: ""
            it.toObject(Offer::class.java)?.copy(id = it.id, jobId = jobId)
        }
    }

    suspend fun updateOfferStatus(jobId: String, offerId: String, status: String) {
        db.collection("jobs").document(jobId)
            .collection("offers").document(offerId)
            .update(mapOf(
                "status" to status,
                "respondedAt" to com.google.firebase.Timestamp.now()
            )).await()
    }

    suspend fun sendCounterOffer(jobId: String, offerId: String, counterPrice: Double, message: String) {
        db.collection("jobs").document(jobId)
            .collection("offers").document(offerId)
            .update(mapOf(
                "status" to "counter_offer",
                "counterPrice" to counterPrice,
                "negotiationMessage" to message,
                "respondedAt" to com.google.firebase.Timestamp.now()
            )).await()
    }

    suspend fun respondToCounterOffer(jobId: String, offerId: String, accept: Boolean) {
        if (accept) {
            db.collection("jobs").document(jobId)
                .collection("offers").document(offerId)
                .update(mapOf(
                    "status" to "accepted",
                    "respondedAt" to com.google.firebase.Timestamp.now()
                )).await()
        } else {
            db.collection("jobs").document(jobId)
                .collection("offers").document(offerId)
                .delete().await()
        }
    }

    // Contractor Profiles
    suspend fun fetchContractorProfile(userId: String): ContractorProfile? {
        val doc = db.collection("contractorProfiles").document(userId).get().await()
        return if (doc.exists()) doc.toObject(ContractorProfile::class.java) else null
    }

    suspend fun fetchAllContractors(): List<ContractorProfile> {
        val snapshot = db.collection("contractorProfiles").get().await()
        return snapshot.documents.mapNotNull { it.toObject(ContractorProfile::class.java) }
    }

    suspend fun updateContractorProfile(profile: ContractorProfile) {
        require(profile.userId.isNotBlank()) { "Invalid user ID" }
        db.collection("contractorProfiles").document(profile.userId).set(profile).await()
    }

    // Completed Jobs
    suspend fun addCompletedJob(completedJob: CompletedJob): String {
        val docRef = db.collection("completedJobs").add(completedJob).await()
        return docRef.id
    }

    suspend fun fetchCompletedJobs(contractorId: String): List<CompletedJob> {
        val snapshot = db.collection("completedJobs")
            .whereEqualTo("contractorId", contractorId)
            .orderBy("completedDate", Query.Direction.DESCENDING)
            .get().await()
        return snapshot.documents.mapNotNull { it.toObject(CompletedJob::class.java)?.copy(id = it.id) }
    }

    // Real-time updates
    fun observeJobs(status: String? = null): Flow<List<Job>> = callbackFlow {
        var query: Query = db.collection("jobs")
        if (status != null) query = query.whereEqualTo("status", status)
        
        val listener = query.orderBy("datePosted", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    close(error)
                    return@addSnapshotListener
                }
                
                val jobs = snapshot?.documents?.mapNotNull { 
                    it.toObject(Job::class.java)?.copy(id = it.id) 
                } ?: emptyList()
                
                trySend(jobs)
            }
        
        awaitClose { listener.remove() }
    }

    fun observeJobsByUser(userId: String): Flow<List<Job>> = callbackFlow {
        val listener = db.collection("jobs")
            .whereEqualTo("createdBy", userId)
            .orderBy("datePosted", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    close(error)
                    return@addSnapshotListener
                }
                
                val jobs = snapshot?.documents?.mapNotNull { 
                    it.toObject(Job::class.java)?.copy(id = it.id) 
                } ?: emptyList()
                
                trySend(jobs)
            }
        
        awaitClose { listener.remove() }
    }

    fun observeOffersForJob(jobId: String): Flow<List<Offer>> = callbackFlow {
        val listener = db.collection("jobs").document(jobId)
            .collection("offers")
            .orderBy("createdAt", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    close(error)
                    return@addSnapshotListener
                }
                
                val offers = snapshot?.documents?.mapNotNull { 
                    it.toObject(Offer::class.java)?.copy(id = it.id, jobId = jobId) 
                } ?: emptyList()
                
                trySend(offers)
            }
        
        awaitClose { listener.remove() }
    }

    fun observeOffersByContractor(contractorId: String): Flow<List<Offer>> = callbackFlow {
        val listener = db.collectionGroup("offers")
            .whereEqualTo("contractorId", contractorId)
            .orderBy("createdAt", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    close(error)
                    return@addSnapshotListener
                }
                
                val offers = snapshot?.documents?.mapNotNull {
                    val jobId = it.reference.parent.parent?.id ?: ""
                    it.toObject(Offer::class.java)?.copy(id = it.id, jobId = jobId)
                } ?: emptyList()
                
                trySend(offers)
            }
        
        awaitClose { listener.remove() }
    }

    fun observeJob(jobId: String): Flow<Job?> = callbackFlow {
        val listener = db.collection("jobs").document(jobId)
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    close(error)
                    return@addSnapshotListener
                }
                
                val job = if (snapshot?.exists() == true) {
                    snapshot.toObject(Job::class.java)?.copy(id = snapshot.id)
                } else {
                    null
                }
                
                trySend(job)
            }
        
        awaitClose { listener.remove() }
    }
}


